### PR TITLE
npm tasks

### PR DESCRIPTION
**Pull request by emmenko**
This supersedes #248 

/cc @gaearon

### CODE REVIEW

**Comment by gaearon** [comment]
How does making a Github release fit into this workflow? Should I just do a release via web interface separately based on the pushed tag?

**Comment by emmenko** [comment]
When you do a release, a github tag is created. Then you need to manually create the release note for that existing tag (but that's always a manual process).

Or do you mean something else?

**Comment by gaearon** [comment]
Yeah this seems right.
Some tools also automate creating release notes so I just wanted to clarify.

This looks great to me.

File: .mversionrc [hunk 1]
```diff
@@ -0,0 +1,9 @@
+{
+  "commitMessage": "Bumps to version v%s",
```
Comment (gaearon) [inline .mversionrc]: Can we just change it to `%s` please

File: package.json [hunk 2]
```diff
@@ -4,14 +4,16 @@
   "description": "Atomic Flux with hot reloading",
   "main": "lib/index.js",
   "scripts": {
-    "browser": "scripts/browser",
```
Comment (gaearon) [inline package.json]: I think `browser` got lost. We still want to have browser builds.

File: package.json [hunk 3]
```diff
@@ -4,14 +4,16 @@
...
-    "build": "scripts/build",
-    "clean": "scripts/clean",
-    "lint": "scripts/lint",
-    "prepublish": "scripts/prepublish",
```
Comment (gaearon) [inline package.json]: I'm still not convinced we need to remove `prepublish`.
I want installing from Git to “just work”. That's what `prepublish` does if I'm not mistaken.

**Comment by emmenko** [comment]
Cool :+1: 

One thing though to remember: the patch/minor/major tasks work for normal semver numbers, so normal lifecycle releases. 
If you want to release alpha/beta/-rc/whatever you should directly use the `mversion` cli (see [examples](https://www.npmjs.com/package/mversion#examples)), or do it manually, just to be sure.

**Comment by gaearon** [comment]
Oh, okay, thanks for explaining.

File: package.json (hunk 3 above)
Comment (emmenko) [inline package.json]: Sure. Should it be `prepublish: npm run build` ?

File: package.json (hunk 2 above)
Comment (emmenko) [inline package.json]: Oops, my bad sorry.

File: .mversionrc (hunk 1 above)
Comment (emmenko) [inline .mversionrc]: Sure

File: package.json (hunk 3 above)
Comment (gaearon) [inline package.json]: Yes I think so. Could also skip linting. I think it's better to have granular “low-level” commands and make some higher level commands on top of them.

File: package.json (hunk 2 above)
Comment (emmenko) [inline package.json]: Should I call it `browser` or `dist`?

File: package.json (hunk 3 above)
Comment (emmenko) [inline package.json]: :+1:

File: package.json (hunk 2 above)
Comment (gaearon) [inline package.json]: `umd` is probably even better

Comment (gaearon) [inline package.json]: or `standalone`. Whatever you like more

Comment (emmenko) [inline package.json]: `umd` it is ;)

**Comment by emmenko** [comment]
@gaearon ok, I made the changes and force pushed. Should be ok now.

File: .mversionrc [hunk 4]
```diff
@@ -0,0 +1,9 @@
+{
+  "commitMessage": "Bumps to version %s",
```
Comment (gaearon) [inline .mversionrc]: just %s please

Comment (emmenko) [inline .mversionrc]: Oh, I thought `vs%` without `v`. Will change it, sorry

File: .mversionrc [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+  "tagName": "v%s",
+  "scripts": {
+    "preupdate": "npm test && npm run umd",
```
Comment (gaearon) [inline .mversionrc]: Why not build it as part of `npm run build`?

File: package.json [hunk 6]
```diff
@@ -4,14 +4,18 @@
...
+    "lint": "eslint src test examples",
+    "prepublish": "npm run build",
+    "release": "mversion patch -m",
+    "release:patch": "mversion patch -m",
+    "release:minor": "mversion minor -m",
+    "release:major": "mversion major -m",
+    "test": "npm run lint && NODE_ENV=test mocha --compilers js:babel/register --recursive",
+    "test:watch": "npm test -- --watch",
+    "test:cov": "npm run lint && babel-node ./node_modules/.bin/isparta cover ./node_modules/.bin/_mocha -- --recursive",
+    "umd": "npm run clean && npm run lint && webpack src/index.js dist/redux.js && NODE_ENV=production webpack src/index.js dist/redux.min.js"
```
Comment (gaearon) [inline package.json]: I don't like that we do `clean` and `lint` in most commands. Isn't it better to have a few low-level “focused” commands that do just one thing (e.g. call `webpack`) and then assemble higher-level commands from them?

For example right now linting and cleaning will happen twice if I build both lib & UMD.

File: .mversionrc (hunk 5 above)
Comment (emmenko) [inline .mversionrc]: The `umd` part? Makes sense!

File: package.json (hunk 2 above)
Comment (gaearon) [inline package.json]: Actually let's call it `build:umd`.

File: package.json (hunk 6 above)
Comment (emmenko) [inline package.json]: I've simplified now. Build will clean and test will lint.

**Comment by emmenko** [comment]
@gaearon feel free to do the final tweaks the way you like it ;)

**Comment by gaearon** [comment]
Sure, thanks! I'll actually use some of these for a boilerplate I'm preparing. (Tired of doing the same setup for every lib.)

File: package.json (hunk 2 above)
Comment (emmenko) [inline package.json]: Done

**Comment by emmenko** [comment]
> Tired of doing the same setup for every lib

I know! :)

PS: Glad to help.

**Comment by gaearon** [comment]
Thanks again! I'll merge tomorrow.
In the meantime, I extracted some stuff into https://github.com/gaearon/library-boilerplate

**Comment by emmenko** [comment]
Good idea! Thanks for the link.

**Comment by valscion** [comment]
Wow this PR clarifies the build process a lot and makes it easier to start hacking on redux. :+1:

**Comment by mikekidder** [comment]
(Window compatibility)
npm tasks that set environment variables for Node, it requires a slight change. 
For example (from/to):

    "build:umd": "webpack src/index.js dist/redux.js && NODE_ENV=production webpack src/index.js dist/redux.min.js"

    "build:umd": "webpack src/index.js dist/redux.js && set NODE_ENV=production&& webpack src/index.js dist/redux.min.js"

I mentioned it in https://github.com/gaearon/library-boilerplate/issues/2.  Should work for Mac as well (I hope) to make it all cross-platform.  The no space after setting variable is the hack discovery that makes it possible in Windows.

**Comment by gaearon** [comment]
@mikekidder This doesn't pass `NODE_ENV=production` to Webpack for some reason.

**Comment by gaearon** [comment]
I tried to fix Windows build in #379.

**Comment by mikekidder** [comment]
Yes, "set" on OS X was questionable, I didn't have a Mac to test.  I'll respond further in #379

**Pull request merged**